### PR TITLE
For 25980: Upgrade navigation, lifecycle, fragment and core to latest versions.

### LINKED ISSUE #25980: Upgrade jetpack navigation & co to latest versions
At the time of opening this issue, the current version for navigation is 2.5.0. We aim to upgrade to 2.5.1, a version that should contain suppression of warnings in the generated safe args code. ( see https://issuetracker.google.com/issues/237725966 ).

Note: As navigation 2.5.0 depends on lifecycle 2.5.0, this issue will involve upgrading `androidx.lifecycle `and possibly other dependencies (core, fragment?)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20975)

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

Fixes #25980

Fixes #25980

### CODE REVIEW

File: buildSrc/src/main/java/Dependencies.kt [hunk 1]
```diff
@@ -21,6 +21,7 @@ object Versions {
     const val jna = "5.8.0"
 
     const val androidx_compose = "1.1.1"
+    const val androidx_compose_compiler = "1.2.0-rc02"
```
Comment (mcarare) [inline buildSrc/src/main/java/Dependencies.kt:24]: Upgrading to 1.2.0 would involve also upgrading Kotlin.

**Comment by JohanLorenzo** [comment]
Hey there! I'm sorry to jump in this PR. I'm just rebasing it because https://github.com/mozilla-mobile/fenix/pull/26073 reached `main`. It renamed the expected check from `pr-complete` to `complete-pr`. This impacts mergify and it may prevent you from merging this PR. That's why the rebase is needed. Let me know if you have any questions.

File: app/build.gradle [hunk 2]
```diff
@@ -449,7 +449,7 @@ configurations {
 
 tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).configureEach {
     kotlinOptions {
-        freeCompilerArgs += "-Xopt-in=kotlinx.coroutines.ExperimentalCoroutinesApi"
+        freeCompilerArgs += "-opt-in=kotlinx.coroutines.ExperimentalCoroutinesApi"
```
Comment (mcarare) [inline app/build.gradle:452]: -Xopt-in is deprecated.

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleSettingsFragment.kt [hunk 3]
```diff
@@ -83,6 +85,10 @@ class LocaleSettingsFragment : Fragment() {
         })
     }
 
+    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
+        TODO("Not yet implemented")
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleSettingsFragment.kt:89]: Can we just return `false` here to indicate a menu item wasn't handled?

File: app/src/main/java/org/mozilla/fenix/settings/logins/fragment/LoginDetailFragment.kt [hunk 4]
```diff
@@ -105,7 +108,6 @@ class LoginDetailFragment : SecureFragment(R.layout.fragment_login_detail) {
 
     override fun onCreate(savedInstanceState: Bundle?) {
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/logins/fragment/LoginDetailFragment.kt:109]: Let's remove `onCreate` if we aren't doing anything with the override

File: buildSrc/src/main/java/Dependencies.kt (hunk 1 above)
Comment (gabrielluong) [inline buildSrc/src/main/java/Dependencies.kt:24]: Can probably drop the rc02 here now :)

Comment (gabrielluong) [reply buildSrc/src/main/java/Dependencies.kt:24]: Curious if we have an issue to upgrade Kotlin next so that we can also bump the Compose version

Comment (mcarare) [reply buildSrc/src/main/java/Dependencies.kt:24]: We have https://github.com/mozilla-mobile/android-components/issues/12489 for AC. It also involves upgrading Room.

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleSettingsFragment.kt (hunk 3 above)
Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleSettingsFragment.kt:89]: Yes, I already did that for `SavedLoginsFragment`, I guess this slipped through.

File: app/src/main/java/org/mozilla/fenix/settings/search/AddSearchEngineFragment.kt [hunk 5]
```diff
@@ -143,7 +147,8 @@ class AddSearchEngineFragment :
 
                 true
             }
-            else -> super.onOptionsItemSelected(item)
+            // other options are not handled by this menu provider
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/search/AddSearchEngineFragment.kt:150]: I want to suggest dropping this comment. It should generally be understood by the `onMenuItemSelected` kDoc that a `false` returned value means the menu item option wasn't handled. I also don't want us to set a precedent where we have to add this comment for all the other `else -> false` in these `onMenuItemSelected`.

Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/settings/search/AddSearchEngineFragment.kt:150]: Sure!

**Comment by Amejia481** [comment]
Same issue as the Focus https://github.com/mozilla-mobile/focus-android/pull/7379#issuecomment-1221094720,
[Locat logs](https://github.com/mozilla-mobile/fenix/files/9390353/fenix.txt). I rebased on top of https://github.com/mozilla-mobile/android-components/pull/12501/.

Tried to load cnn.com and the app crashed 💥 .

```
2022-08-21 20:36:15.380 17574-17574/org.mozilla.focus.debug D/CompatibilityChangeReporter: Compat change id reported: 171979766; UID 10342; state: ENABLED
2022-08-21 20:36:15.502 17166-17292/org.mozilla.focus.debug W/Web Content: [JavaScript Warning: "Use of the motion sensor is deprecated." {file: "https://registry.api.cnn.io/bundles/fave/theoplayer-9be11a9d/theoplayer" line: 42}]
2022-08-21 20:36:15.520 17062-17107/org.mozilla.focus.debug D/CompatibilityChangeReporter: Compat change id reported: 136069189; UID 10342; state: ENABLED
2022-08-21 20:36:15.520 17062-17107/org.mozilla.focus.debug W/System.err: java.lang.SecurityException: To use the sampling rate of 0 microseconds, app needs to declare the normal permission HIGH_SAMPLING_RATE_SENSORS.
2022-08-21 20:36:15.521 17062-17107/org.mozilla.focus.debug W/System.err:     at android.hardware.SystemSensorManager$BaseEventQueue.enableSensor(SystemSensorManager.java:792)
2022-08-21 20:36:15.521 17062-17107/org.mozilla.focus.debug W/System.err:     at android.hardware.SystemSensorManager$BaseEventQueue.addSensor(SystemSensorManager.java:712)
2022-08-21 20:36:15.521 17062-17107/org.mozilla.focus.debug W/System.err:     at android.hardware.SystemSensorManager.registerListenerImpl(SystemSensorManager.java:209)
2022-08-21 20:36:15.521 17062-17107/org.mozilla.focus.debug W/System.err:     at android.hardware.SensorManager.registerListener(SensorManager.java:832)
2022-08-21 20:36:15.521 17062-17107/org.mozilla.focus.debug W/System.err:     at android.hardware.SensorManager.registerListener(SensorManager.java:739)
2022-08-21 20:36:15.521 17062-17107/org.mozilla.focus.debug W/System.err:     at org.mozilla.gecko.GeckoAppShell.enableSensor(GeckoAppShell.java:636)
2022-08-21 20:36:15.521 17062-17107/org.mozilla.focus.debug W/System.err:     at org.mozilla.gecko.mozglue.GeckoLoader.nativeRun(Native Method)
2022-08-21 20:36:15.521 17062-17107/org.mozilla.focus.debug W/System.err:     at org.mozilla.gecko.GeckoThread.run(GeckoThread.java:677)
```

**Comment by mcarare** [comment]
The app also crashes on main, with Pixel 4/ Android 13 when opening cnn.com.

**Comment by fice-t** [comment]
@mcarare It should be noted that updating androidx.core to 1.8.0 also fixes #26674 and #26808 as I mentioned in #27259 (didn't notice this issue updated core... 😅).

For reference, see this bug that was fixed in core 1.8.0-alpha04: https://issuetracker.google.com/issues/180881870

File: app/src/main/java/org/mozilla/fenix/settings/search/AddSearchEngineFragment.kt (hunk 5 above)
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/settings/search/AddSearchEngineFragment.kt:150]: Bumping this as I get back to reviewing and testing this

**Review by jonalmeida — APPROVED** [review]

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragment.kt [hunk 6]
```diff
@@ -134,6 +136,9 @@ class BookmarkFragment : LibraryPageFragment<BookmarkNode>(), UserInteractionHan
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
+
+        requireActivity().addMenuProvider(this, viewLifecycleOwner, Lifecycle.State.RESUMED)
```
[the remark this answers is not in the document]
Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragment.kt:140]: I tested both places and saw no visual difference. It should work either way, as the view and viewLifecycleOwner are accessible in both lifecycle states. Also, the menu becomes visible only in Fragment's `onResume()` (set by the third param `Lifecycle.State.RESUMED`. Choosing ` Lifecycle.State.STARTED` can cause weird overlapping menus on fragment transitions. ). So I preferred to keep the place exhibited in the [documentation](https://developer.android.com/jetpack/androidx/releases/activity#1.4.0-alpha01).

**Comment by mcarare** [comment]
@jonalmeida just a heads up: after your review and rebasing, I added this [commit](https://github.com/mozilla-mobile/fenix/pull/25991/commits/e56e09f240c48dc13364b662540947e1e0636df1) (so you can review it if needed).

**Comment by jonalmeida** [comment]
> @jonalmeida just a heads up: after your review and rebasing, I added this [commit](https://github.com/mozilla-mobile/fenix/pull/25991/commits/e56e09f240c48dc13364b662540947e1e0636df1) (so you can review it if needed).

All good - feel free to still land it. :)

**Pull request merged**